core/src/main/java/org/elasticsearch/tasks/Task.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.action.admin.cluster.node.tasks.list.TaskInfo;
+import org.elasticsearch.common.inject.Provider;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Current task information
+ */
+public class Task {
+
+    private final long id;
+
+    private final String type;
+
+    private final String action;
+
+    private final Provider<String> description;
+
+    private final AtomicBoolean canceled;
+
+    private Listener listener;
```

Comment:
should this listener be volatile? I also wonder if we should fail / throw an exceptin if the listener is non-null when we try to set it? -- there is a nice class in lucene called `org.apache.lucene.util.SetOnce<Listener>` which does that for you